### DISCUSSION TITLE
Как создавать подпись УКЭП (CADESCOM_CADES_X_LONG_TYPE_1)

### DISCUSSION THREAD

**Discussion by firsovdevelop** [CryptoPro/libcore#24] [General]
Нам нужно реализовать создание и валидацию подписи по стандарту CADESCOM_CADES_X_LONG_TYPE_1 (Усиленная квалифицированная) с штампом времени. В примерах не показано как это можно сделать. А код примеров к КриптоПро.NET и corefx не подходит (нет объекта CPSigner, например).
Подскажите, пожалуйста, как используя эту библиотеку libcore сделать валидную подпись УКЭП?

**Comment by Fasjeit** [comment]
Для исходных билиотеках дотнета есть поддержка только SignedCms. Дорабатывали Cms только в плане поддержки ГОСТов, XLT1 отдельно не поддерживали.

Сформировать XLT1 в настоящий момент можно только через P/Invoke CAdES.

**Reply by firsovdevelop** [reply]
Скажите, пожалуйста, значит и валидировать подпись XLT1 с помощью данной библиотеки пока не получится?

**Reply by Fasjeit** [reply]
Да, в настоящий момент только P/Invoke

**Reply by ghost** [reply]
Подскажите пжл, есть ли пример проекта на .net core, где бы использовалась либа cades/cadescom через p/Invoke? Или, может, где-то в одном месте есть информация, где брать нужную dll, .h файл к ней, и, желательно, описание функций и используемых параметров?

**Reply by Fasjeit** [reply]
@egnp Если у вас windows то самый простой способ использование CadesCom.

Необходимо установить [sdk](https://www.cryptopro.ru/products/cades/sdk), подключить библиотеку Crypto-Pro CAdES 1.0 Type Library (в студии add reference - com - Crypto-Pro CAdES 1.0 Type Library).

Далее можно пользоваться объектами из неё для выполнения криптографических операций.
Документация - https://cpdn.cryptopro.ru/default.asp?url=content/cades/cades-com-usage.html

Для си документация рядом - https://cpdn.cryptopro.ru/default.asp?url=content/cades/cades-c-usage.html

Про p/Invoke, если нужно, можно почитать на сайте ms - https://learn.microsoft.com/ru-ru/dotnet/standard/native-interop/pinvoke
